### PR TITLE
add quoting logic of deactivated, stand-by, close-only modes of vaults

### PR DESCRIPTION
**Pull request by tqin7**
### Changelist
in deactivated and stand-by mode, don't quote
in close-only mode: quote only if short or long

in another PR, i'm thinking:
1. when setting a vault to deactivated or stand-by, cancel its existing orders, if any

### Test Plan
unit tests

### Author/Reviewer Checklist
- [ ] If this PR has changes that result in a different app state given the same prior state and transaction list, manually add the `state-breaking` label.
- [ ] If the PR has breaking postgres changes to the indexer add the `indexer-postgres-breaking` label.
- [ ] If this PR isn't state-breaking but has changes that modify behavior in `PrepareProposal` or `ProcessProposal`, manually add the label `proposal-breaking`.
- [ ] If this PR is one of many that implement a specific feature, manually label them all `feature:[feature-name]`.
- [ ] If you wish to for mergify-bot to automatically create a PR to backport your change to a release branch, manually add the label `backport/[branch-name]`.
- [ ] Manually add any of the following labels: `refactor`, `chore`, `bug`.

## Summary by CodeRabbit

- **New Features**
	- Enhanced vault order management with improved status checks and order placement logic.
	- Introduced a new cancellation function for vault orders to streamline operations.
	- Updated function to retrieve both vault and quoting parameters for better clarity.

- **Bug Fixes**
	- Refined logic to prevent unnecessary order processing in inactive vault states.

- **Documentation**
	- Updated function names and return values for clarity regarding vault and quoting parameters.

- **Tests**
	- Expanded and restructured test cases to validate new functionalities and edge cases for vault orders and parameters.

### CODE REVIEW

File: protocol/x/vault/keeper/orders.go [hunk 1]
```diff
@@ -344,6 +367,26 @@ func (k Keeper) GetVaultClobOrders(
...
+		reduceOnlyMaxOrderSize.Quo(reduceOnlyMaxOrderSize, stepSize)
+		reduceOnlyMaxOrderSize.Mul(reduceOnlyMaxOrderSize, stepSize)
+		// If vault is long, only need sell orders (even indices).
+		i := 0
+		if leverage.Sign() < 0 {
+			// If vault is short, only need buy orders (odd indices).
+			i = 1
+		}
+		for ; i < len(orders); i += 2 {
+			reduceOnlyOrders[i/2] = orders[i]
```
Comment (vincentwschau) [inline protocol/x/vault/keeper/orders.go:384]: nit: Rather than do the division, maybe just keep another index counter that you increment per loop iteration?

File: protocol/x/vault/keeper/orders.go [hunk 2]
```diff
@@ -344,6 +367,26 @@ func (k Keeper) GetVaultClobOrders(
...
 	}
 
+	if vaultParams.Status == types.VaultStatus_VAULT_STATUS_CLOSE_ONLY {
+		// In close-only mode with non-zero leverage.
+		reduceOnlyOrders := make([]*clobtypes.Order, quotingParams.Layers)
+		reduceOnlyMaxOrderSize := k.GetVaultInventoryInPerpetual(ctx, vaultId, perpetual.Params.Id)
+		stepSize := lib.BigU(clobPair.StepBaseQuantums)
+		reduceOnlyMaxOrderSize.Quo(reduceOnlyMaxOrderSize, stepSize)
+		reduceOnlyMaxOrderSize.Mul(reduceOnlyMaxOrderSize, stepSize)
+		// If vault is long, only need sell orders (even indices).
```
Comment (vincentwschau) [inline protocol/x/vault/keeper/orders.go:377]: non-blocking: I wonder if this code could be built slightly more future-proof by setting some `closeOnlySide` based on long/short, then filtering out the orders from the generated `orders` rather than using even/odd indices.

File: protocol/x/vault/keeper/orders_test.go [hunk 3]
```diff
@@ -218,127 +286,464 @@ func TestRefreshAllVaultOrders(t *testing.T) {
...
+					clientIdIsCanonical: []bool{
+						false,
+						false,
+						false,
+						false,
+					},
+				},
+			},
+		},
+		"Success - Orders refresh due to status changing to close-only. No more orders": {
```
Comment (vincentwschau) [inline protocol/x/vault/keeper/orders_test.go:566]: nit: Note that the vault has no position in this test-case?

File: protocol/x/vault/keeper/orders_test.go [hunk 4]
```diff
@@ -218,127 +286,464 @@ func TestRefreshAllVaultOrders(t *testing.T) {
...
+						)
 					},
-				)
+					ordersRefreshed:     []bool{}, // no orders
+					orderSides:          []clobtypes.Order_Side{},
+					clientIdIsCanonical: []bool{},
+				},
+			},
+		},
+		"Success - Orders refresh due to status changing to close-only. Sell orders only": {
```
Comment (vincentwschau) [inline protocol/x/vault/keeper/orders_test.go:604]: nit: Another similar test-case for `STAND_BY` and `DEACTIVATED`?

File: protocol/x/vault/keeper/orders_test.go [hunk 5]
```diff
@@ -797,7 +1271,143 @@ func TestGetVaultClobOrders(t *testing.T) {
...
+			// order_size = 10% * 1250 / 3000 ~= 0.04166666667
+			// order_size_base_quantums = 0.04166666667e9 ~= 41_666_667
+			// round down to nearest multiple of step_base_quantums=1_000.
+			expectedOrderQuantums: []uint64{
+				41_666_000,
+				41_666_000,
+				41_666_000,
+			},
+		},
+		"Success - Vault Clob 1, close-only status, 2 layers, leverage 0.6, sell orders only, order size": {
```
Comment (vincentwschau) [inline protocol/x/vault/keeper/orders_test.go:1344]: nit: Update title to indicate this is for capped order size = position size.

File: protocol/x/vault/keeper/orders.go [hunk 6]
```diff
@@ -344,6 +367,26 @@ func (k Keeper) GetVaultClobOrders(
 		orders[2*i+1] = constructOrder(clobtypes.Order_SIDE_BUY, i, orderIds[2*i+1])
 	}
 
+	if vaultParams.Status == types.VaultStatus_VAULT_STATUS_CLOSE_ONLY {
+		// In close-only mode with non-zero leverage.
+		reduceOnlyOrders := make([]*clobtypes.Order, quotingParams.Layers)
+		reduceOnlyMaxOrderSize := k.GetVaultInventoryInPerpetual(ctx, vaultId, perpetual.Params.Id)
+		stepSize := lib.BigU(clobPair.StepBaseQuantums)
+		reduceOnlyMaxOrderSize.Quo(reduceOnlyMaxOrderSize, stepSize)
```
Comment (vincentwschau) [inline protocol/x/vault/keeper/orders.go:375]: nit: Return no orders if this is 0?

File: protocol/x/vault/keeper/orders.go (hunk 1 above)
Comment (tqin7) [reply protocol/x/vault/keeper/orders.go:384]: see below comment

File: protocol/x/vault/keeper/orders_test.go (hunk 4 above)
Comment (tqin7) [reply protocol/x/vault/keeper/orders_test.go:604]: right now we simply skip `refresh orders` if status is `stand_by` or `deactivated`. I'm planning to have another change that cancels existing orders if status is changed to `stand_by` or `deactivated` and will add these test cases in that change

**Review by vincentwschau — APPROVED** [review]
